File: 数仓表结构/asset_package/src/main/java/com/weshare/interfaces/impl/ProjectInfoImpl.java
package com.weshare.interfaces.impl;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.weshare.entity.ProjectInfo;
import com.weshare.interfaces.AssetFiles;
import org.apache.spark.api.java.function.MapFunction;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Encoders;
import org.apache.spark.sql.SparkSession;

import java.io.Serializable;

/**
 * Created by mouzwang on 2020-11-30 15:42
 */
public class ProjectInfoImpl implements AssetFiles, Serializable {
    private static final long serialVersionUID = 1L;

    public Dataset<ProjectInfo> process(Dataset<String> dataset) {
        Dataset<ProjectInfo> mapedDS = dataset.map(new MapFunction<String, ProjectInfo>() {
            @Override
            public ProjectInfo call(String value) throws Exception {
                ProjectInfo projectInfo = new ProjectInfo();
                JsonObject jsonObject = new JsonParser().parse(value).getAsJsonObject();
                projectInfo.setProjectId(jsonObject.get("project_id").getAsString());
                projectInfo.setProjectName(jsonObject.get("project_name").getAsString());
                projectInfo.setProjectStage(jsonObject.get("project_stage").getAsString());
                projectInfo.setAssetSide(jsonObject.get("asset_side").getAsString());
                projectInfo.setFundSide(jsonObject.get("fund_side").getAsString());
                projectInfo.setYear(jsonObject.get("year").getAsString());
                projectInfo.setTerm(jsonObject.get("term").getAsString());
                projectInfo.setRemarks(jsonObject.get("remarks") == null ? null :
                        jsonObject.get("remarks").getAsString());
                projectInfo.setProjectFullName(jsonObject.get("project_full_name").getAsString());
                projectInfo.setAssetType(jsonObject.get("asset_type").getAsString());
                projectInfo.setProjectType(jsonObject.get("project_type").getAsString());
                projectInfo.setMode(jsonObject.get("mode").getAsString());
                projectInfo.setProjectTime(jsonObject.get("project_time").getAsString());
                projectInfo.setProjectBeginDate(jsonObject.get("project_begin_date") == null ?
                        null : jsonObject.get("project_begin_date").getAsString());
                projectInfo.setProjectEndDate(jsonObject.get("project_end_date") == null ?
                        null : jsonObject.get("project_end_date").getAsString());
                projectInfo.setAssetPoolType(jsonObject.get("asset_pool_type").getAsString());
                projectInfo.setPublicOffer(jsonObject.get("public_offer") == null ? null : jsonObject.get("public_offer").getAsString());
                projectInfo.setDataSource(jsonObject.get("data_source").getAsString());
                projectInfo.setCreateUser(jsonObject.get("create_user") == null ? null : jsonObject.get("create_user").getAsString());
                projectInfo.setCreateTime(jsonObject.get("create_time").getAsString());
                projectInfo.setUpdateTime(jsonObject.get("update_time").getAsString());
                return projectInfo;
            }
        }, Encoders.bean(ProjectInfo.class));
        return mapedDS;
    }


    @Override
    public void insertData(Dataset<String> dataset, SparkSession sparkSession, String fileId) {
        Dataset<ProjectInfo> processDS = process(dataset);
        processDS.registerTempTable("temp_project_info");
        sparkSession.sql("insert overwrite table dim.project_info partition (project_id = '" + fileId + "') " +
                "select " +
                "projectName,projectStage,assetSide,fundSide,year,term,remarks,projectFullName," +
                "assetType,projectType,mode,projectTime,projectBeginDate,projectEndDate," +
                "assetPoolType,publicOffer,dataSource,createUser,createTime,updateTime " +
                "from temp_project_info ");
    }

    @Override
    public void updateData(Dataset<String> dataset, SparkSession sparkSession, String fileId) {
        Dataset<ProjectInfo> processDS = process(dataset);
        processDS.registerTempTable("temp_project_info");
        sparkSession.sql("insert overwrite table dim.project_info partition (project_id='" + fileId +
                "') " +
                "select " +
                "projectName,projectStage,assetSide,fundSide,year,term,remarks,projectFullName," +
                "assetType,projectType,mode,projectTime,projectBeginDate,projectEndDate," +
                "assetPoolType,publicOffer,dataSource,createUser,createTime,updateTime " +
                "from temp_project_info ");
    }

    @Override
    public void deleteData(Dataset<String> dataset, SparkSession sparkSession, String fileId, String importId) {
        sparkSession.sql("alter table dim.project_info drop partition (project_id = '" + fileId + "') ");
        sparkSession.sql("alter table dim.project_due_bill_no drop partition (project_id = '" + fileId + "') ");
    }
}

File: 数仓表结构/asset_package/src/main/scala/com/weshare/interface/impl/ProjectInfoImpl.scala
package com.weshare.interface.impl

import com.weshare.interface.AssetFile
import org.apache.spark.sql.functions.lit

/**
  * @author ximing.wei 2021-05-27 14:50:07
  */
class ProjectInfoImpl extends AssetFile {

    override def insertData(): Unit = {
        logger.info(logBase, "执行 插入 语句！")

        logger.info(logBase, "将获取的数据，按字段创建临时表  project_info_tmp ！")
        data.select(
            lit("project_id"),
            lit("project_name"),
            lit("project_stage"),
            lit("asset_side"),
            lit("fund_side"),
            lit("year"),
            lit("term"),
            lit("remarks"),
            lit("project_full_name"),
            lit("asset_type"),
            lit("project_type"),
            lit("mode"),
            lit("project_time"),
            lit("project_begin_date"),
            lit("project_end_date"),
            lit("asset_pool_type"),
            lit("public_offer"),
            lit("data_source"),
            lit("create_user"),
            lit("create_time"),
            lit("update_time")
        ).createOrReplaceTempView("project_info_tmp")

        logger.info(logBase, "执行 插入 语句 HQL！")
        sparkSession.sql(
            """
              |insert overwrite table dim.project_info partition(project_id)
              |select
              |  project_name,
              |  project_stage,
              |  asset_side,
              |  fund_side,
              |  year,
              |  term,
              |  remarks,
              |  project_full_name,
              |  asset_type,
              |  project_type,
              |  mode,
              |  project_time,
              |  project_begin_date,
              |  project_end_date,
              |  asset_pool_type,
              |  public_offer,
              |  data_source,
              |  create_user,
              |  create_time,
              |  update_time,
              |  project_id
              |from project_info_tmp
            """.stripMargin)
        logger.info(logBase, "插入 语句执行 完成！")
    }

    override def updateData(): Unit = {
        logger.info(logBase, "执行 更新 语句！")

        logger.info(logBase, "调用 插入 语句！")
        insertData()
        logger.info(logBase, "更新 语句执行 完成！")
    }

    override def deleteData(): Unit = {
        logger.info(logBase, "执行 删除 语句！")

        logger.info(logBase, "对于表 project_info 不应该存在 删除 操作！并抛出异常！")
        throw new IllegalArgumentException("对于表 project_info 不应该存在 删除 操作！")
    }
}
